### ISSUE TITLE
[v2] Introduce a `ManagedResourceDefinition` type

### ISSUE THREAD

**Issue by negz** [crossplane/crossplane#2262]
### What problem are you facing?

An unintended (but nice) property of Crossplane's `CompositeResourceDefinition` (XRD) type is that it gives us a Crossplane specific artifact representing the existence of a Crossplane specific class of resource (i.e. Composite Resources or XRs). This means it's possible for a subject to walk the tree of Crossplane resources from a `Configuration` to a `ConfigurationRevision` to a `CompositeResourceDefinition` to the XRs defined by that XRD without ever needing to be granted RBAC access to anything but Crossplane-specific types. In practice an XRD is a proxy for a CRD, but Crossplane users can get all the information they need from the XRD without ever needing RBAC access to read CRDs.

Managed resources (and provider configs) are a different story. In order to walk the tree from a provider to a particular managed resource you need to traverse CRDs - e.g. `Provider` to `ProviderRevision` to `CustomResourceDefinition` to the managed resources defined by that CRD. This implies that browsing the Crossplane Resource Model (XRM) requires access to get and list CRDs. In any particular Kubernetes cluster running Crossplane it's likely that some subset of the CRDs don't have anything to do with Crossplane, and that the cluster operator might be reluctant to grant access to _all_ CRDs when in fact they really only want to enable a subject to map a particular `ProviderRevision` to the managed resources and provider configs it delivers.

### How could Crossplane help solve your problem?

Crossplane could support specific `ManagedResourceDefinition` and `ProviderConfigDefinition` types, which would be proxies for CRDs. This would allow cluster operators to grant access to the `ManagedResourceDefinition` type (for example) to allow a subject to discover what Crossplane managed resources exist without necessarily having access to discover _all_ custom resources that exist.

**Comment by negz** [comment]
This proposal has a bunch of implications, e.g.:

* Would we have a provider package up a bunch of Managed Resource Definitions (MRDs) instead of CRDs?
* Would we need a controller that created a CRD for every MRD it saw?
* Would we need to make controller-tools generate MRDs from Crossplane types, instead of CRDs?

**Comment by muvaf** [comment]
I'm trying to gauge whether it's worth taking on the added mental complexity of additional proxy types. Is the following list of benefits for MRDs complete?
* Instead of `list, get all CRDs` RBAC rules, they will be able to work with `list, get all MRDs`.
* When the explorers tries to figure out how to query instances of managed resources, they can go from `Provider -> ProviderRevision -> MRDs -> their instances` instead of `Provider -> ProviderRevision -> ALL CRDs -> their instances`.

Regarding RBAC, I don't feel like people would be uncomfortable giving read-only rights for CRDs (not their instances). Do you have some example scenarios where read-only access to all CRDs could be harmful?

> This would allow cluster operators to grant access to the ManagedResourceDefinition type (for example) to allow a subject to discover what Crossplane managed resources exist without necessarily having access to discover all custom resources that exist.

I think they can filter CRDs by `spec.names.categories`, which has `managed` or we can label managed resource CRDs with all of their categories (`aws`, `managed`, `providerconfig`, `crossplane`) so they can query CRDs with those labels to get the list of managed resource or provider config CRDs. Is there a reason we need type-level differentiation rather than labeling, other than RBAC point above?

**Comment by negz** [comment]
> Is there a reason we need type-level differentiation rather than labeling, other than RBAC point above?

It's mostly RBAC. The only other potential pros I could think of are:

* Having a nice tightly scoped, Crossplane specific proxy type is arguably a better user experience when trying to list _types_ (not instances) than trying to filter CRDs.
* It could be a place to put guard rails on provider opinions; e.g. I could imagine either the package manager or an admission control webhook ensuring that managed resources were 'compliant' - e.g. had the schema bits we expect an MR to have.
* The symmetry of `Config -> Revision -> XRD -> CRD -> XR` and `Provider -> Revision -> MRD -> CRD -> MR` is a bit nice?

**Comment by muvaf** [comment]
We got to discuss this with @negz and wanted to write down the `Lock` idea; other than listing CRDs with specific labels, package manager can store the list of CRDs in the `Lock` object similar to how it stores `Provider` resources it installed.

**Comment by negz** [comment]
Dropping another note here - while there are many other ways to skin this cat, I could see an MRD being a good place to add additional Crossplane specific metadata about a type - for example mapping a kind like `MySQLInstance` to a human readable friendly name like "MySQL Instance".

**Comment by bassam** [comment]
/refresh I believe this is still relevant

**Comment by negz** [comment]
I've been thinking about this as I work on the Crossplane v2 prototype (https://github.com/crossplane/crossplane/pull/6255). I'm starting to cool on the idea.

The idea of a ManagedResourceDefinition type occurred to me while working on https://github.com/upbound/xgql. xgql models relationships between types. That's why I noticed you could track a Configuration -> ConfigurationRevision -> XRD -> XR using only Crossplane-specific types.

It's true that you can't track a Provider -> MRs without going via the CustomResourceDefinitions the Provider installs. In retrospect I'm not sure that's a real problem though. Especially if you're not trying to build something like xgql.

My understanding is that you're not _really_ supposed to use CRDs to discover what types exist anyway. You're supposed to use the [Discovery API](https://kubernetes.io/docs/concepts/overview/kubernetes-api/#discovery-api). Today you can browse the discovery API and look for resources in the categories `crossplane` and `managed` to discover types of MRs. Admittedly this doesn't tell you what provider they're from.

I think the main value to users of an MRD would be giving us somewhere to record what connection secrets exist.

If it was easy to introduce the MRD I'd probably lean toward doing so, even just for the symmetry with XRDs. I don't think it'd be easy to introduce, though.

It'd be easy to add an MRD type to Crossplane and add a controller that reconciles MRDs by creating CRDs. What's harder is making providers generate MRDs instead of CRDs. Today providers use https://github.com/kubernetes-sigs/controller-tools to generate CRD manifests from Go types. This is a well-known pattern in the Kubernetes ecosystem.

If we wanted to generate MRDs instead we'd either need to:

* Fork and maintain a variant of controller-tools that generated MRDs instead
* Generate CRDs and post-process them by changing them to an MRD (and injecting connection secret information)

Post-processing would probably be the easier option. We'd need to make sure controller-tools never generated CRDs with fields that don't exist in the MRD type though, including in older versions of Crossplane.

**Comment by negz** [comment]
> I think the main value to users of an MRD would be giving us somewhere to record what connection secrets exist.

We could do this without adding a new type by using annotations on MR CRDs. For example:

```yaml
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  name: instances.rds.aws.upbound.io
  annotations:
    cd.crossplane.io/username: "The master username for the database."
    cd.crossplane.io/master_user_secret: "A block that specifies the master user secret. Only available when manage_master_user_password is set to true."
spec:
  ...
```

Annotation keys are limited to 63 characters, while values are limited to 253 characters.

We'd still need to inject these annotations on the generated CRDs - presumably as a post-processing step.

**Comment by negz** [comment]
> Today you can browse the discovery API and look for resources in the categories crossplane and managed to discover types of MRs. Admittedly this doesn't tell you what provider they're from.

We could address this by having the package manager explicitly add categories to CRDs to create a relationship between the provider and a resource. For example by adding `crossplane`, `aws`, and `provider-aws` categories to any CRD installed by a Provider named `provider-aws.`

**Comment by turkenh** [comment]
> For example by adding crossplane, aws, and provider-aws categories to any CRD installed by a Provider named provider-aws.

FWIW, we are [already](https://github.com/crossplane-contrib/provider-upjet-aws/blob/main/package/crds/accessanalyzer.aws.upbound.io_analyzers.yaml#L14) very close to this:

```
    categories:
    - crossplane
    - managed
    - aws
```

**Comment by negz** [comment]
> FWIW, we are [already](https://github.com/crossplane-contrib/provider-upjet-aws/blob/main/package/crds/accessanalyzer.aws.upbound.io_analyzers.yaml#L14) very close to this:

Yeah I know. I'd like the package manager to enforce that though. Today it's up to provider authors [^1] to remember to set those categories. I'd rather make sure the categories are always there in a predictable format even if the provider author forgets to include them.


[^1]: Though [upjet does it for you]( https://github.com/crossplane/upjet/blob/72675757bb33ae6153813b0f6d1d0bf7bba9dba3/pkg/pipeline/templates/crd_types.go.tmpl#L50), so it'd only ever be an potential issue for hand written providers.

**Comment by negz** [comment]
I've now flip flopped and am planning to stick with this for v2. There's a few other places I think it could be useful. One is that it gives us a pretty clean way to enable/disable CRDs, by toggling whether the MRD is rendered out to a CRD or not.

**Comment by jbw976** [comment]
I'm going to go ahead and close this feature as complete for core crossplane v2.0, and https://github.com/crossplane/crossplane/issues/6611 can follow on through the provider ecosystem on its own timeline ✅
